From 4124d7f4b8ef56bfac3b0969788612c0e21719bb Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Tue, 10 Aug 2021 10:12:34 +0300
Subject: [PATCH] [SW-50880] gaudi2: arc: run pcie initialization on arc0

This patch add support to perform PCIe initialization on secondary
CPU ARC0 and let primary CPU ARC1 continue with preboot. Upon successful
initialization of PCIe, secondary CPU ARC0 will go in sleep mode or halt
mode which will be indication in preboot and PCIe EROM for PCIe is
initialized.

Change-Id: I6cb7bce55308fd2e05adfe37557d9359544e08f1
---

diff --git a/app/ext_rom/Makefile b/app/ext_rom/Makefile
index 01639f9..757e907 100644
--- a/app/ext_rom/Makefile
+++ b/app/ext_rom/Makefile
@@ -30,11 +30,13 @@
 	$(PREBOOT_ROOT_DIR)/lib/common/pll.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/pci.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/uart.o \
+	$(PREBOOT_ROOT_DIR)/lib/common/boot_if.o \
 	$(PREBOOT_ROOT_DIR)/app/ppboot/ppboot.o \
 	$(HL_DEV_ARCH)/$(HL_TARGET).o \
 	$(HL_DEV_ARCH)/../common/$(HL_TARGET).o \
 	$(HL_DEV_ARCH)/../common/common.o \
-	$(HL_DEV_ARCH)/common.o
+	$(HL_DEV_ARCH)/common.o \
+	$(HL_DEV_ARCH)/secondary.o
 
 PREBOOT_SRCS_C := $(shell ls $(PREBOOT_ROOT_DIR)/lib/common/*.c)
 PREBOOT_SRCS_C += $(shell ls $(PREBOOT_ROOT_DIR)/lib/preboot/*.c)
diff --git a/app/ext_rom/erom_pcie.c b/app/ext_rom/erom_pcie.c
index 861a4f4..1cbaa59 100644
--- a/app/ext_rom/erom_pcie.c
+++ b/app/ext_rom/erom_pcie.c
@@ -19,6 +19,14 @@
 	udelay_init();
 
 	hl_uart_init();
+	init_secondary();
 	hl_pcie_init();
-}
 
+	/*
+	 * since pcie initialization is performed by secondary
+	 * ARC (ARC0), wait here till it finishes pcie link
+	 * initialization.
+	 */
+	while (!is_pcie_initialized())
+		;
+}
diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 38d3443..b4cffe7 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -81,7 +81,7 @@
 	regmap_writel(0x0, mmPSOC_GLOBAL_CONF_BOOT_SEQ_EXT_LD);
 }
 
-static void init_secondary(void)
+void init_secondary(void)
 {
 	halt_n_reset_arc(0);
 
@@ -251,3 +251,104 @@
 	*loadaddr = copy_preboot();
 	return true;
 }
+
+static void preboot_pcie_monitoring(void)
+{
+	/* Wait for PRSTn */
+	while (!regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN))
+		;
+
+	/* Only after PRSTn is de-asserted */
+	preboot_pcie_enable();
+
+	while (1) {
+		unsigned int prstn, ltssm, smlh, rdlh;
+
+		prstn = regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN);
+		ltssm = regmap_readl(mmPCIE_AUX_LTSSM_EN);
+		smlh = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
+		rdlh = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
+
+		if (prstn && smlh && rdlh)
+			break;
+
+		/*
+		 * Some host assert prstn during link up while moving to GEN3
+		 * This cause BAR size to be reset
+		 * We stop link up process by disabling ltssm
+		 * till BAR is reconfigured
+		 */
+		if (!prstn && ltssm) {
+			regmap_writel(0, mmPCIE_AUX_LTSSM_EN);
+			continue;
+		}
+
+		if (prstn && !ltssm) {
+			regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
+			continue;
+		}
+	}
+
+	/* W/A H/W bug that may lead to have a wrong mask (0x1) in LTSSM */
+	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
+}
+
+void secondary_main(void)
+{
+	preboot_pcie_monitoring();
+
+	/* Put secondary CPU in sleep state until further notice */
+	while (1)
+		__asm__ __volatile__("sleep\n\t" : : );
+}
+
+void preboot_pcie_wait_for_link(void)
+{
+	u32 *secondary_iccm_addr = (u32 *) ARC0_ICCM_BASE;
+
+	/* init dev id as soon as possible, while waiting for prstn */
+	preboot_device_id_init();
+
+	/*
+	 * update secondary entry address to 0th location of ICCM
+	 * ARC once put in running state will jump to address
+	 * specified in 0th location of ICCM
+	 */
+	*secondary_iccm_addr = (u32) &secondary_entry;
+	run_arc(0);
+}
+
+bool is_pcie_initialized(void)
+{
+	u32 secondary_sts;
+
+	secondary_sts = regmap_readl(mmPSOC_ARC0_CFG_RUN_HALT_ACK);
+
+	/*
+	 * Secondary CPU is put to sleep once PCIe initialization
+	 * is done. So notify PCIe is initialized if secondary CPU
+	 * (ARC0) is in sleep state or in halt state.
+	 */
+	if ((secondary_sts & PSOC_ARC0_CFG_RUN_HALT_ACK_SYS_SLEEP_R_MASK) ||
+		(secondary_sts & PSOC_ARC0_CFG_RUN_HALT_ACK_SYS_HALT_R_MASK))
+		return true;
+	else
+		return false;
+}
+
+void preboot_secondary_cpu_stop(void)
+{
+	u32 *secondary_iccm_addr = (u32 *) ARC0_ICCM_BASE;
+
+	/*
+	 * Secondary CPU will be in sleep once pcie initialization
+	 * is done, halt and reset ARC0 so that it can ready
+	 * to load and run zephyr images received via FIT.
+	 */
+	*secondary_iccm_addr = 0x0;
+
+	halt_n_reset_arc(0);
+
+	regmap_writel(ARC_RST_VEC_ADDR, mmPSOC_ARC0_CFG_RST_VEC_ADDR);
+	regmap_writel(ARC_RST_VALUE, mmPSOC_ARC0_CFG_ARC_RST);
+}
diff --git a/arch/gaudi2/arc/common.h b/arch/gaudi2/arc/common.h
index 7b43c90..4c951b3 100644
--- a/arch/gaudi2/arc/common.h
+++ b/arch/gaudi2/arc/common.h
@@ -20,6 +20,7 @@
 #define FW_SRAM_BASE		0x27FD0000
 #define SP_SRAM_BASE		0x27FE0000
 #define ARC_ICCM_BASE		0x27040000
+#define ARC0_ICCM_BASE		0x27000000
 #define SP_SRAM_PREBOOT_OFFSET	0x1000
 #define PPBOOT_TEXT_BASE	(FW_SRAM_BASE + 0xC)
 #define PREBOOT_TEXT_BASE	(SP_SRAM_BASE + SP_SRAM_PREBOOT_OFFSET + 0x8)
@@ -34,7 +35,7 @@
  * Let PPBoot start from the offset of 0x402C, for using DCCM
  * for its own BSS and Stack.
  */
-#define PPBOOT_DCCM_BASE	0x8000402C
+#define PPBOOT_DCCM_BASE	0x2706402C
 
 #define SDRAM_BASE		0x11000000
 #define HL_SPI_BASE		0x26000000
@@ -57,4 +58,9 @@
 
 #define FW_WORD_NUM		0x2200	/* FW_SIZE / 2 */
 
+#ifndef __ASSEMBLY__
+void init_secondary(void);
+extern void secondary_entry(void);
+#endif
+
 #endif /* COMMON_GAUDI2_ARC_GNU_H_ */
diff --git a/arch/gaudi2/arc/secondary.S b/arch/gaudi2/arc/secondary.S
new file mode 100644
index 0000000..753af2c
--- /dev/null
+++ b/arch/gaudi2/arc/secondary.S
@@ -0,0 +1,12 @@
+/* SPDX-License-Identifier:     GPL-2.0+ */
+/*
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+.global secondary_entry
+secondary_entry:
+	mov	sp, 0x27020400	/* stack top set to secondary CPU DCCM 1k offset */
+	mov	fp, sp
+	bl	secondary_main
diff --git a/arch/gaudi2/arm/common.c b/arch/gaudi2/arm/common.c
index 13411d3..f780dfa 100644
--- a/arch/gaudi2/arm/common.c
+++ b/arch/gaudi2/arm/common.c
@@ -48,6 +48,7 @@
 #define ARC1_DCCM_BASE_ADDR		0x1000007FF7060000ull
 #define ARC1_DCCM_SIZE			0x8000			/* 32KB */
 
+#define HL_PCIE_LINK_RETRIES		50
 
 u64 lkd_dev_spec_get_img_addr(uintptr_t base)
 {
@@ -127,41 +128,6 @@
 	void (*image_entry)(void) = (void *)loadaddr;
 
 	image_entry();
-}
-
-static void run_arc(u8 id)
-{
-	u32 run_ack_offset, run_req_offset;
-	u32 run_ack_mask;
-	u64 start;
-
-	if (id) {
-		run_ack_offset = mmPSOC_ARC1_CFG_RUN_HALT_ACK;
-		run_req_offset = mmPSOC_ARC1_CFG_RUN_HALT_REQ;
-		run_ack_mask = PSOC_ARC1_CFG_RUN_HALT_ACK_RUN_ACK_MASK;
-
-	} else {
-		run_ack_offset = mmPSOC_ARC0_CFG_RUN_HALT_ACK;
-		run_req_offset = mmPSOC_ARC0_CFG_RUN_HALT_REQ;
-		run_ack_mask = PSOC_ARC0_CFG_RUN_HALT_ACK_RUN_ACK_MASK;
-
-	}
-
-	/* Set ARC in running state */
-	regmap_writel(BIT(0), run_req_offset);
-
-	start = timestamp_get_us();
-	while (!(regmap_readl(run_ack_offset) & run_ack_mask)) {
-
-		if (timeout_expired(start, ARC_RUN_ACK_TIMEOUT_US)) {
-			preboot_puts_hex("No running state ack from ARC ",
-					id, false);
-			break;
-		}
-	}
-
-	regmap_writel(0x0, run_req_offset);
-
 }
 
 static void init_arc(void)
@@ -318,3 +284,66 @@
 {
 	return (u32)mmPCIE_DBI_BASE & 0xFFFFFFF;
 }
+
+void preboot_pcie_wait_for_link(void)
+{
+	u8 retry = HL_PCIE_LINK_RETRIES;
+
+	/* init dev id as soon as possible, while waiting for prstn */
+	preboot_device_id_init();
+
+	/* Wait for PRSTn */
+	while (!regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN))
+		;
+
+	/* Only after PRSTn is de-asserted */
+	preboot_pcie_enable();
+
+	preboot_puts("Waiting for PCIe UP status\n");
+
+	while (1) {
+		unsigned int prstn, ltssm, smlh, rdlh;
+
+		prstn = regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN);
+		ltssm = regmap_readl(mmPCIE_AUX_LTSSM_EN);
+		smlh = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
+		rdlh = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
+
+		if (prstn && smlh && rdlh) {
+			preboot_puts("PCIe link is UP!\n");
+			break;
+		}
+
+		/*
+		 * Do not wait indefinitely on PCIe link up status
+		 * Wait for 5 seconds (50 retries with 100 milli-
+		 * seconds sleep), else bailout.
+		 */
+		retry--;
+		if (!retry) {
+			preboot_puts("Timed-out waiting for PCIe link UP!\n");
+			break;
+		}
+
+		udelay(100000);
+
+		/*
+		 * Some host assert prstn during link up while moving to GEN3
+		 * This cause BAR size to be reset
+		 * We stop link up process by disabling ltssm
+		 * till BAR is reconfigured
+		 */
+		if (!prstn && ltssm) {
+			regmap_writel(0, mmPCIE_AUX_LTSSM_EN);
+			continue;
+		}
+
+		if (prstn && !ltssm) {
+			regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
+			continue;
+		}
+	}
+
+	/* W/A H/W bug that may lead to have a wrong mask (0x1) in LTSSM */
+	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
+}
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index f79ea66..0aa5c3c 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -51,12 +51,12 @@
 
 #define PCIE_LTR_DISABLE		0
 #define PCIE_CAP_HDR_NXT_OFFSET		0x1f8
-#define HL_PCIE_LINK_RETRIES             50
 
 #define DUMP_SEL_SRNS_MASK		0x1
 #define DUMP_SEL_SSC_MASK		0x2
 
 #define ARC_HALT_ACK_RETRIES		100
+#define ARC_RUN_ACK_TIMEOUT_US		100
 
 void lkd_dev_spec_regs(struct cpu_dyn_regs *regs)
 {
@@ -413,68 +413,6 @@
 	/* Enable and Mask LTSSM (erratum 102) */
 	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
 }
-void preboot_pcie_wait_for_link(void)
-{
-	u8 retry = HL_PCIE_LINK_RETRIES;
-
-	/* init dev id as soon as possible, while waiting for prstn */
-	preboot_device_id_init();
-
-	/* Wait for PRSTn */
-	while (!regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN))
-		;
-
-	/* Only after PRSTn is de-asserted */
-	preboot_pcie_enable();
-
-	preboot_puts("Waiting for PCIe UP status\n");
-
-	while (1) {
-		unsigned int prstn, ltssm, smlh, rdlh;
-
-		prstn = regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN);
-		ltssm = regmap_readl(mmPCIE_AUX_LTSSM_EN);
-		smlh = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
-		rdlh = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
-
-		if (prstn && smlh && rdlh) {
-			preboot_puts("PCIe link is UP!\n");
-			break;
-		}
-
-		/*
-		 * Do not wait indefinitely on PCIe link up status
-		 * Wait for 5 seconds (50 retries with 100 milli-
-		 * seconds sleep), else bailout.
-		 */
-		retry--;
-		if (!retry) {
-			preboot_puts("Timed-out waiting for PCIe link UP!\n");
-			break;
-		}
-
-		udelay(100000);
-
-		/*
-		 * Some host assert prstn during link up while moving to GEN3
-		 * This cause BAR size to be reset
-		 * We stop link up process by disabling ltssm
-		 * till BAR is reconfigured
-		 */
-		if (!prstn && ltssm) {
-			regmap_writel(0, mmPCIE_AUX_LTSSM_EN);
-			continue;
-		}
-
-		if (prstn && !ltssm) {
-			regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
-			continue;
-		}
-	}
-
-	/* W/A H/W bug that may lead to have a wrong mask (0x1) in LTSSM */
-	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
-}
 
 void preboot_delay(void) {}
 
@@ -652,3 +590,47 @@
 	regmap_writel(0x1, rst_offset);
 }
 
+/*
+ * Set ARC in running state
+ * 1. Set run bit in RUN_HALT_REQ register
+ * 2. Wait for ACK from core in RUN_HALT_ACK register
+ * 3. Clear run bit in RUN_HALT_REQ register
+ */
+void run_arc(u8 id)
+{
+	u32 run_ack_offset, run_req_offset;
+	u32 run_ack_mask;
+	u64 start;
+	u32 cpu_boot_err_val;
+
+	if (id) {
+		run_ack_offset = mmPSOC_ARC1_CFG_RUN_HALT_ACK;
+		run_req_offset = mmPSOC_ARC1_CFG_RUN_HALT_REQ;
+		run_ack_mask = PSOC_ARC1_CFG_RUN_HALT_ACK_RUN_ACK_MASK;
+
+	} else {
+		run_ack_offset = mmPSOC_ARC0_CFG_RUN_HALT_ACK;
+		run_req_offset = mmPSOC_ARC0_CFG_RUN_HALT_REQ;
+		run_ack_mask = PSOC_ARC0_CFG_RUN_HALT_ACK_RUN_ACK_MASK;
+
+	}
+
+	/* Set ARC in running state */
+	regmap_writel(BIT(0), run_req_offset);
+
+	start = timestamp_get_us();
+	while (!(regmap_readl(run_ack_offset) & run_ack_mask)) {
+
+		if (timeout_expired(start, ARC_RUN_ACK_TIMEOUT_US)) {
+			preboot_puts_hex("No running state ack from ARC ",
+					id, false);
+			cpu_boot_err_val = (!id) ?
+				CPU_BOOT_ERR0_ARC0_RUN_ACK_NOT_RCVD :
+				CPU_BOOT_ERR0_ARC1_RUN_ACK_NOT_RCVD;
+			cpu_boot_err_clrset(cpu_boot_err_val, true);
+			break;
+		}
+	}
+
+	regmap_writel(0x0, run_req_offset);
+}
diff --git a/arch/gaudi2/common/hl.h b/arch/gaudi2/common/hl.h
index e721776..04f0cf1 100644
--- a/arch/gaudi2/common/hl.h
+++ b/arch/gaudi2/common/hl.h
@@ -21,5 +21,6 @@
 void set_pci_msix_pba_offset_val(void);
 void set_slv_cfg_reg_val(void);
 void halt_n_reset_arc(u8 id);
+void run_arc(u8 id);
 
 #endif /* ARCH_GAUDI2_COMMON_HL_H_ */
